### ISSUE TITLE
amended commit can cause scm feature to fail

### ISSUE THREAD

**Issue by DavidZemon** [conan-io/conan#4465]
Conan clients are 1.11.2 inside Ubuntu 18.10 Docker images with Python 2

We have CI server with two slaves and our own Artifactory instance where Conan packages are uploaded at the end of a build.

The following error as occurred

```
[10:07:24]	++ /start.sh conan install /home/redlionci/tcAgentHome/work/35034cfe92273263 --install-folder /home/redlionci/tcAgentHome/work/35034cfe92273263/build/Release --build missing --build outdated --update
[10:07:24]	Configuration:
[10:07:24]	[settings]
[10:07:24]	arch=armv7hf
[10:07:24]	arch_build=x86_64
[10:07:24]	build_type=Release
[10:07:24]	compiler=gcc
[10:07:24]	compiler.libcxx=libstdc++11
[10:07:24]	compiler.version=7.3
[10:07:24]	os_build=Linux
[10:07:24]	[options]
[10:07:24]	[build_requires]
[10:07:24]	*: gcc-arm-none-eabi/7.3.1@wsbu/stable
[10:07:24]	[env]
[10:07:24]	
[10:07:25]	st-hal-provider/1.0.0@wsbu/stable: Package is up to date
[10:07:25]	argtable3/3.0.4-1@wsbu/testing: WARN: Package is corrupted, removing folder: /home/captain/.conan/data/argtable3/3.0.4-1/wsbu/testing/package/868ef901de8d5dc2677c480b7dd3c17eb1caa7b6
[10:07:25]	argtable3/3.0.4-1@wsbu/testing: Outdated package!
[10:07:25]	FreeRTOS/10.1.1@wsbu/stable: Package is up to date
[10:07:25]	OpenOCD/0.10.0+10+20181020+0522@wsbu/stable: Package is up to date
[10:07:26]	gcc-arm-none-eabi/7.3.1@wsbu/stable: Package is up to date
[10:07:26]	gcc-arm-none-eabi/7.3.1@wsbu/stable: Package is up to date
[10:07:26]	PROJECT: Installing /home/redlionci/tcAgentHome/work/35034cfe92273263/conanfile.py
[10:07:26]	Requirements
[10:07:26]	    FreeRTOS/10.1.1@wsbu/stable from local cache - No remote
[10:07:26]	    argtable3/3.0.4-1@wsbu/testing from local cache - No remote
[10:07:26]	    st-hal-provider/1.0.0@wsbu/stable from local cache - No remote
[10:07:26]	Packages
[10:07:26]	    FreeRTOS/10.1.1@wsbu/stable:5ab84d6acfe1f23c4fae0ab88f26e3a396351ac9 - Cache
[10:07:26]	    argtable3/3.0.4-1@wsbu/testing:868ef901de8d5dc2677c480b7dd3c17eb1caa7b6 - Build
[10:07:26]	    st-hal-provider/1.0.0@wsbu/stable:4a557f9b673226121f5b69b60ec6e36aafbbb9f4 - Cache
[10:07:26]	Build requirements
[10:07:26]	    OpenOCD/0.10.0+10+20181020+0522@wsbu/stable from local cache - No remote
[10:07:26]	    gcc-arm-none-eabi/7.3.1@wsbu/stable from local cache - No remote
[10:07:26]	Build requirements packages
[10:07:26]	    OpenOCD/0.10.0+10+20181020+0522@wsbu/stable:44fcf6b9a7fb86b2586303e3db40189d3b511830 - Cache
[10:07:26]	    gcc-arm-none-eabi/7.3.1@wsbu/stable:44fcf6b9a7fb86b2586303e3db40189d3b511830 - Cache
[10:07:26]	
[10:07:26]	Cross-build from 'Linux:x86_64' to 'None:armv7hf'
[10:07:26]	FreeRTOS/10.1.1@wsbu/stable: Already installed!
[10:07:26]	OpenOCD/0.10.0+10+20181020+0522@wsbu/stable: Already installed!
[10:07:26]	gcc-arm-none-eabi/7.3.1@wsbu/stable: Already installed!
[10:07:26]	st-hal-provider/1.0.0@wsbu/stable: Already installed!
[10:07:26]	argtable3/3.0.4-1@wsbu/testing: Applying build-requirement: gcc-arm-none-eabi/7.3.1@wsbu/stable
[10:07:26]	argtable3/3.0.4-1@wsbu/testing: WARN: Build folder is dirty, removing it: /home/captain/.conan/data/argtable3/3.0.4-1/wsbu/testing/build/868ef901de8d5dc2677c480b7dd3c17eb1caa7b6
[10:07:26]	argtable3/3.0.4-1@wsbu/testing: Building your package in /home/captain/.conan/data/argtable3/3.0.4-1/wsbu/testing/build/868ef901de8d5dc2677c480b7dd3c17eb1caa7b6
[10:07:26]	argtable3/3.0.4-1@wsbu/testing: WARN: Trying to remove corrupted source folder
[10:07:26]	argtable3/3.0.4-1@wsbu/testing: WARN: This can take a while for big packages
[10:07:26]	argtable3/3.0.4-1@wsbu/testing: Configuring sources in /home/captain/.conan/data/argtable3/3.0.4-1/wsbu/testing/source
[10:07:26]	argtable3/3.0.4-1@wsbu/testing: Getting sources from url: 'https://github.com/wsbu/argtable3.git'
[10:07:26]	/bin/sh: warning: setlocale: LC_ALL: cannot change locale (en_US.UTF-8)
[10:07:26]	Cloning into '.'...
[10:07:27]	/bin/sh: warning: setlocale: LC_ALL: cannot change locale (en_US.UTF-8)
[10:07:27]	/bin/sh: warning: setlocale: LC_ALL: cannot change locale (en_US.UTF-8)
[10:07:27]	/bin/sh: warning: setlocale: LC_ALL: cannot change locale (en_US.UTF-8)
[10:07:27]	fatal: reference is not a tree: 9745c5e8e039a717da4172b6dbfb481f4c1ff51a
[10:07:27]	ERROR: Command 'git checkout "9745c5e8e039a717da4172b6dbfb481f4c1ff51a"' returned non-zero exit status 128
[10:07:27]	Process exited with code 1
[10:07:27]	Process exited with code 1 (Step: Prepare Conan (Command Line))
[10:07:28]	Step Prepare Conan (Command Line) failed
```

I _think_ the following steps caused the problem, but don't have the time to try and reproduce it on purpose.

1. Build a library using the SCM attribute like so on CI slave 1 and upload to artifactory:
   ```
   scm = {
       'type': 'git',
       'url': 'auto',
       'revision': 'auto'
    }
   ```
2. Make changes to the library and use `git commit --amend`
3. Build the library on CI slave 2 and upload to artifactory
4. Build an application which uses the library on CI slave 1

The recipe in Artifactory contains the correct revision:

```
scm = {"revision": "4dcce447b1883f37e6b8ce84703376bf3fb73095",
    "type": "git",
    "url": "https://github.com/wsbu/argtable3.git"}
```

but it looks like the new recipe isn't actually being downloaded on CI slave 1, despite the log showing that Conan _knows_ it is outdated.

**Comment by lasote** [comment]
Could you confirm that locally, in the build that failed, the recipe contains the wrong commit?
I'm not sure how to face this issue.

**Comment by DavidZemon** [comment]
In the Conan cache on CI slave 1 the recipe contains the wrong commit, yes.

**Comment by lasote** [comment]
And the `--update` doesn't work?

**Comment by DavidZemon** [comment]
Nope. already in use and double checked by looking at the first line of the console output in the code block above

**Comment by lasote** [comment]
I saw in your ouput log:

```
[10:07:26]	Requirements
[10:07:26]	    FreeRTOS/10.1.1@wsbu/stable from local cache - No remote
[10:07:26]	    argtable3/3.0.4-1@wsbu/testing from local cache - No remote
[10:07:26]	    st-hal-provider/1.0.0@wsbu/stable from local cache - No remote
```
It looks like the references have no remote associated. That's why it can't update.
You can try to specify a remote with `-r`, I think forcing the remote the update will work.

**Comment by DavidZemon** [comment]
Was there a bug at some point that did not set remotes after a `conan upload` command? I'm trying to figure out how these packages could have gotten into the local cache without having a remote defined.

**Comment by lasote** [comment]
The remote is set when downloading the packages I would say. If you are installing the packages in a regular way the remote should be associates. Could be a bug, yes, but hard to know.

**Comment by DavidZemon** [comment]
I've been getting this problem a lot over the last week. Here's a snippet from a `conan install ...` command:

```
[21:01:12]	googletest/1.8.0@wsbu/stable: Package is up to date
[21:01:12]	wsbu-docgen/0.1.3@wsbu/stable: Package is up to date
[21:01:14]	catch2/2.5.0@bincrafters/stable: WARN: Package catch2/bincrafters is being deprecated. Change yours to require catchorg instead
[21:01:14]	wsbu-docgen/0.1.3@wsbu/stable: Package is up to date
[21:01:15]	catch2/2.5.0@bincrafters/stable: Package is up to date
[21:01:15]	Version ranges solved
[21:01:15]	    Version range '^2.6.2' required by 'None' resolved to 'ddio/2.6.4@wsbu/stable'
[21:01:15]	    Version range '^0.1.0' required by 'None' resolved to 'wsbu-docgen/0.1.3@wsbu/stable'
[21:01:15]	    Version range '^2.4.0' required by 'None' resolved to 'catch2/2.5.0@bincrafters/stable'
[21:01:15]	    Version range '^0.1.0' required by 'None' resolved to 'wsbu-docgen/0.1.3@wsbu/stable'
[21:01:15]	
[21:01:15]	PROJECT: Installing /home/redlionci/tcAgentHome/work/dcd10b52160da67e/conanfile.py
[21:01:15]	Requirements
[21:01:15]	    ddio/2.6.4@wsbu/stable from local cache - No remote
[21:01:15]	Packages
[21:01:15]	    ddio/2.6.4@wsbu/stable:160b4bac5177b0f9a5f4857d00317fe0862a8a02 - Build
[21:01:15]	Build requirements
[21:01:15]	    catch2/2.5.0@bincrafters/stable from 'conan-center' - Cache
[21:01:15]	    googletest/1.8.0@wsbu/stable from local cache - No remote
[21:01:15]	    wsbu-docgen/0.1.3@wsbu/stable from local cache - No remote
[21:01:15]	Build requirements packages
[21:01:15]	    catch2/2.5.0@bincrafters/stable:5ab84d6acfe1f23c4fae0ab88f26e3a396351ac9 - Cache
[21:01:15]	    googletest/1.8.0@wsbu/stable:9bfdcfa2bb925892ecf42e2a018a3f3529826676 - Cache
[21:01:15]	    wsbu-docgen/0.1.3@wsbu/stable:5ab84d6acfe1f23c4fae0ab88f26e3a396351ac9 - Cache
```

And yet, the `registry.json` file shows that googletest has a defined remote. Is there anywhere else Conan could be getting remote information?

```
redlionci@ci1:~/.conan$ grep googletest registry.json 
  "googletest/1.8.0@wsbu/stable:ec36c36799b7c64ac87634a31cfe9aeae27f47fd": "ci",
  "googletest/1.8.0@wsbu/stable:41773acee08366dc867f4471801ee690d0af3d48": "ci",
  "googletest/1.8.0@wsbu/stable:69f10d1e6d78c1962fbe8fb80440f32b1fdf8423": "ci",
  "googletest/1.8.0@wsbu/stable:816eaae1c2a7c47f13ce581f3ded7fb53fda4238": "ci",
  "googletest/1.8.0@wsbu/stable:9bfdcfa2bb925892ecf42e2a018a3f3529826676": "ci",
  "googletest/1.8.0@wsbu/stable:18308414964ec52f799c72a10c5a844d086274cb": "ci",
  "googletest/1.8.0@wsbu/stable:160b4bac5177b0f9a5f4857d00317fe0862a8a02": "ci",
  "googletest/1.8.0@wsbu/stable:89bbc176999b31cf8d641dbc5f6c25d9935600e9": "ci",
```

**Comment by lasote** [comment]
These are the registry entries for the binaries, but what about the recipe? Why is not there? Where that recipe came from?

**Comment by DavidZemon** [comment]
Is this related? Seems odd that `conan info` would show a remote is connected, but `conan install` says no remote

```
captain@d41405c940b5:/home/redlionci/tcAgentHome/work/3ef198c6893c5ec9/bin$ conan info iodb/2.2.0@wsbu/stable
iodb/2.2.0@wsbu/stable
    ID: 937bd37fb52e282e5b4d17239b360b47e35f9c60
    BuildID: None
    Remote: None
    URL: https://bitbucket.org/redlionstl/iodb.git
    License: Proprietary
    Recipe: Cache
    Binary: Cache
    Binary remote: ci
    Creation date: 2019-02-26 18:43:44
    Required by:
        PROJECT
captain@d41405c940b5:/home/redlionci/tcAgentHome/work/3ef198c6893c5ec9/bin$ conan install .. --update --build missing --build outdated
Configuration:
[settings]
arch=x86_64
arch_build=x86_64
build_type=Release
compiler=gcc
compiler.libcxx=libstdc++11
compiler.version=7
os=Linux
os_build=Linux
platform=x86_64
[options]
[build_requires]
[env]

cJSON/1.7.0@wsbu/stable: Package is up to date
iodb/2.2.0@wsbu/stable: WARN: Current package is newer than remote upstream one
iodb/2.2.0@wsbu/stable: Package is up to date
googletest/1.8.1@wsbu/stable: Package is up to date
wsbu-docgen/0.1.3@wsbu/stable: Package is up to date
Version ranges solved
    Version range '^0.1.0' required by 'None' resolved to 'wsbu-docgen/0.1.3@wsbu/stable'

PROJECT: Installing /home/redlionci/tcAgentHome/work/3ef198c6893c5ec9/conanfile.py
Requirements
    cJSON/1.7.0@wsbu/stable from local cache - No remote
    iodb/2.2.0@wsbu/stable from local cache - No remote
Packages
    cJSON/1.7.0@wsbu/stable:5c2ff7bdda7070327f02edf85f4b84eb01fea9bc - Cache
    iodb/2.2.0@wsbu/stable:937bd37fb52e282e5b4d17239b360b47e35f9c60 - Cache
Build requirements
    googletest/1.8.1@wsbu/stable from local cache - No remote
    wsbu-docgen/0.1.3@wsbu/stable from local cache - No remote
Build requirements packages
    googletest/1.8.1@wsbu/stable:9bfdcfa2bb925892ecf42e2a018a3f3529826676 - Cache
    wsbu-docgen/0.1.3@wsbu/stable:5ab84d6acfe1f23c4fae0ab88f26e3a396351ac9 - Cache

cJSON/1.7.0@wsbu/stable: Already installed!
googletest/1.8.1@wsbu/stable: Already installed!
iodb/2.2.0@wsbu/stable: Already installed!
wsbu-docgen/0.1.3@wsbu/stable: Already installed!
PROJECT: Applying build-requirement: googletest/1.8.1@wsbu/stable
PROJECT: Applying build-requirement: wsbu-docgen/0.1.3@wsbu/stable
PROJECT: Generator cmake created conanbuildinfo.cmake
PROJECT: Generator txt created conanbuildinfo.txt
PROJECT: Generated conaninfo.txt
PROJECT: Generated graphinfo
captain@d41405c940b5:/home/redlionci/tcAgentHome/work/3ef198c6893c5ec9/bin$ 
```

**Comment by memsharded** [comment]
The ``scm`` got a replacement in 1.X and was removed in 2.0, closing.
